src/alchemlyb/tests/test_ti_gaussian_quadrature.py

```diff
@@ -0,0 +1,227 @@
+"""Tests for all TI-based estimators in ``alchemlyb``.
+
+"""
+import pandas as pd
+import numpy as np
+import pytest
+
+import alchemlyb
+from alchemlyb.estimators import TI_GQ, TI
+from alchemlyb.parsing import amber
+
+
+@pytest.fixture
+def ethanol_Coulomb(gmx_ethanol_Coulomb_dHdl):
+    dHdl = alchemlyb.concat(gmx_ethanol_Coulomb_dHdl)
+    return dHdl
+
+
+@pytest.fixture
+def ethanol_VDW(gmx_ethanol_VDW_dHdl):
+    dHdl = alchemlyb.concat(gmx_ethanol_VDW_dHdl)
+    return dHdl
+
+
+@pytest.fixture
+def ethanol(ethanol_Coulomb, ethanol_VDW):
+    dHdl = alchemlyb.concat([ethanol_Coulomb, ethanol_VDW])
```

Comment:
Sorry, it could be that I'm doing something wrong, I'm testing this on a set that has two lambda values, such as the ethanol that you have uploaded.

Do you mind check how do I generate the estimate for this? It seems that on my computer

TI().fit(ethanol) gives different result compared with TI_GQ().fit(ethanol)

Reply:
Thanks for the comment. For the ethanol (concat the coulomb and VDW) case, I think TI would take an extra difference between the last state of coulomb `(0.9531, 0.0)` and the first state of vdw `(1.0, 0.0254)`. I'm not sure if this is supposed to happen, as 1.0 is not part of the gaussian quadrature points. While if we use TI_GQ, the coulomb and vdw are calculated separately.
I also tried TI().fit(ethanol_Coulomb) and TI.fit(ethanol_VDW), separately. The results between TI and TI_GQ are more different on ethanol_Coulomb data than on ethanol_VDW data. This might because the quadrature points I used for the simulations are only 5 `(0.0469, 0.2308, 0.5000, 0.7692, 0.9531)` for Coulomb vs 7 `(0.0254, 0.1292, 0.2971, 0.5000, 0.7029, 0.8708, 0.9746)` for VDW. The less lambdas we use, the further away we get from the end points, for example, 0.0254 is closer to 0.0 than 0.0469, and 0.9746 is closer to 1.0 than 0.9531. If we use TI().fit(), usually simulations at 0.0 and 1.0 are expected or extrapolation might be needed, so estimation with TI() on this dataset might cause some inaccuracy and closer to 0.0 and 1.0 is more accurate. This is a difference between TI and TI_GQ. TI_GQ do not need simulations on 0.0 and 1.0.

Reply:
I have generate the same ethanol in water test with the lambds of

```
      coul     vdw
0   0.0000  0.0000
1   0.0469  0.0000
2   0.2308  0.0000
3   0.5000  0.0000
4   0.7692  0.0000
5   0.9531  0.0000
6   1.0000  0.0000
7   1.0000  0.0254
8   1.0000  0.1292
9   1.0000  0.2971
10  1.0000  0.5000
11  1.0000  0.7029
12  1.0000  0.8708
13  1.0000  0.9746
14  1.0000  1.0000
```

I think the ideal case would be that if you run TI and TI_GQ on this set, you would get the same result.
[Archive.zip](https://github.com/alchemistry/alchemlyb/files/11423369/Archive.zip)

Reply:
Thanks for the files! Yes, that's correct. TI would consider all lambda points, while TI_GQ only counts for those between 0 and 1 if they are in the special_points list, but they have different weights for each lambda. The results should be similar between them when there are enough lambda points. I saw a ~0.2 kcal/mol difference in this case between TI and TI_GQ. My guess is that if we increase the number of lambda points the difference would be smaller. I'm trying 12 lambdas for coul and vdw, respectively, but I didn't include the (0.0000  0.0000), (1.0000  0.0000), and (1.0000  1.0000)

Reply:
@xiki-tempula I'm attaching the new dataset of ethanol in water. 12 gaussian quadrature points for coulomb and vdw respectively, plus 3 end states (0.0, 0.0), (1.0, 0.0), (1.0, 1.0). Could you please try it out when you get a chance?

The differences between TI and TI_GQ is ~0.058 kcal/mol in energy and ~0.00039 kcal/mol in error. The energy estimated by MBAR is slightly closer to TI_GQ. I guess more lambda points would bring the results from different methods closer to each other.

[ethanol.zip](https://github.com/alchemistry/alchemlyb/files/11445506/ethanol.zip)

Reply:
@hl2500 Thank you very much for the waiting. I have tried your [ethanol.zip](https://github.com/alchemistry/alchemlyb/files/11445506/ethanol.zip) and I have two comments.

1. > I kind of want all estimator to retain the same interface. I wonder if it is possible to have lambda 0 and lambda 1 in `delta_f_` and `d_delta_f_`. The lambda 0 shall be quite simple which shall be zero. The lambda 1 could just be the same as the last value. Then add a note to the estimator saying that lambda 1 is just the same as the last lambda window.

2. I'm a bit confused as to the result that I got from this ethanol dataset. I tried
3.

Reply:
Thanks for the comments. Unlike the TI method, the values in the table are accumulative sum from one state (row: i) to another (column: j) :
```math
\sum_{k=i}^j \left(wight_k * \left< dHdl  \right>_k \right)
```
rather than the difference between the two states, so the diagonal values are not zeros and lower triangular values are zeros. I thought about putting the values with the opposite sign from the upper triangle, but they are not meaningful values. Or could I make the table symmetrical (the lower triangle is the same as the upper one)?

Reply:
@hl2500 Thanks for the explanation. 
So in all other estimators (TI, MBAR, BAR), the values in the table are accumulative sum from one state (row: i) to another (column: j), so I kind of want to have a similar interface across all estimators.

> I thought about putting the values with the opposite sign from the upper triangle, but they are not meaningful values.

I do agree that it is not a meaningful value but I think putting the values with the opposite sign from the upper triangle might make it more similar to the other estimators.

Reply:
@xiki-tempula Thank you for the suggestions! 

I have entered the values into the lower triangle.

I'm wondering if we could keep the lambdas without adding the 0 and 1? If the simulations were not performed at 0 and 1, the lambdas 0 and 1 would not show in the tables from other estimators as well. For multi-lambda case, we also need to insert intermediate states, such as (1.0, 0.0) or (1.0, 1.0, 0.0). This is a little complicated as the number of lambda types and the number of values for each lambda type vary from different simulations.

Also, do you think I need to upload this ethanol test to alchemtest to replace the previous one with fewer lamndas? Could I change the name of this estimator to TIGQ?

Thank you for your time into this!

Reply:
@xiki-tempula Thank you for the suggestions. I have added the two end states and the new ethanol test has been uploaded, but it seemed that the checks have not been initiated yet.

Reply:
Thanks for addressing the comments. The test is running now.